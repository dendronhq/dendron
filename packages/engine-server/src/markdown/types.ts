import {
  DendronConfig,
  DNoteRefLink,
  DNoteRefLinkRaw,
  DVault,
  NoteProps,
} from "@dendronhq/common-all";
import { Heading, Parent, Root } from "mdast";
import { Processor } from "unified";
import { DendronPubOpts } from "./remark/dendronPub";
import { WikiLinksOpts } from "./remark/wikiLinks";
export { Node as UnistNode } from "unist";
export { VFile } from "vfile";
export { Processor };

// --- General

export type DendronASTRoot = Root & {
  children: DendronASTNode;
};

export type WikiLinkProps = {
  alias: string;
  value: string;
  anchorHeader?: string;
};

export type DendronASTNode = Parent & {
  notes?: NoteProps[];
  children?: Parent["children"] | DendronASTNode[];
};

export enum DendronASTTypes {
  WIKI_LINK = "wikiLink",
  REF_LINK = "refLink",
  REF_LINK_V2 = "refLinkV2",
  BLOCK_ANCHOR = "blockAnchor",
  // Not dendron-specific, included here for convenience
  ROOT = "root",
  HEADING = "heading",
  LIST = "list",
  LIST_ITEM = "listItem",
  PARAGRAPH = "paragraph",
  TABLE = "table",
  TABLE_ROW = "tableRow",
  TABLE_CELL = "tableCell",
  IMAGE = "image",
  FRONTMATTER = "yaml",
  TEXT = "text",
}

export enum DendronASTDest {
  MD_ENHANCED_PREVIEW = "MD_ENHANCED_PREVIEW",
  MD_REGULAR = "MD_REGULAR",
  MD_DENDRON = "MD_DENDRON",
  HTML = "HTML",
}

export enum VaultMissingBehavior {
  FALLBACK_TO_ORIGINAL_VAULT,
  THROW_ERROR,
}

export type DendronASTData = {
  dest: DendronASTDest;
  vault: DVault;
  fname: string;
  wikiLinkOpts?: WikiLinksOpts;
  config: DendronConfig;
  overrides?: Partial<DendronPubOpts>;
  shouldApplyPublishRules?: boolean;
  /**
   * Inidicate that we are currently inside a note ref
   */
  insideNoteRef?: boolean;
};

// --- NODES

export type WikiLinkNoteV4 = DendronASTNode & {
  type: DendronASTTypes.WIKI_LINK;
  value: string;
  data: WikiLinkDataV4;
};

export type WikiLinkDataV4 = {
  alias: string;
  anchorHeader?: string;
  prefix?: string;
  vaultName?: string;
};

export type NoteRefNoteV4_LEGACY = DendronASTNode & {
  type: DendronASTTypes.REF_LINK;
  value: string;
  data: NoteRefDataV4_LEGACY;
};

export type NoteRefNoteV4 = Omit<DendronASTNode, "children"> & {
  type: DendronASTTypes.REF_LINK_V2;
  value: string;
  data: NoteRefDataV4;
};

export type NoteRefNoteRawV4 = Omit<DendronASTNode, "children"> & {
  type: DendronASTTypes.REF_LINK_V2;
  value: string;
  data: NoteRefDataRawV4;
};

export type NoteRefDataV4 = {
  link: DNoteRefLink;
  vaultName?: string;
};

export type NoteRefDataRawV4 = {
  link: DNoteRefLinkRaw;
  vaultName?: string;
};

export type NoteRefDataV4_LEGACY = {
  link: DNoteRefLink;
};

export type BlockAnchor = DendronASTNode & {
  type: DendronASTTypes.BLOCK_ANCHOR;
  id: string;
};

export type Anchor = BlockAnchor | Heading;
