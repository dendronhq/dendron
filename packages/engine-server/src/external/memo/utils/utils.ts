/* eslint-disable no-cond-assign */
/* eslint-disable no-restricted-syntax */
/* eslint-disable no-use-before-define */
import { sort as sortPaths } from "cross-path-sort";
import _ from "lodash";
import path from "path";
import { URI } from "vscode-uri";
import { RefT, WorkspaceCache } from "../types";

export { sortPaths };

const workspaceCache: WorkspaceCache = {
  imageUris: [],
  markdownUris: [],
  otherUris: [],
  allUris: [],
  danglingRefsByFsPath: {},
  danglingRefs: [],
};

const markdownExtRegex = /\.md$/i;
export const REGEX_FENCED_CODE_BLOCK =
  /^( {0,3}|\t)```[^`\r\n]*$[\w\W]+?^( {0,3}|\t)``` *$/gm;

export const containsMarkdownExt = (pathParam: string): boolean =>
  !!markdownExtRegex.exec(path.parse(pathParam).ext);

export const refPattern = "(\\[\\[)([^\\[\\]]+?)(\\]\\])";

// === Utils

export const findUriByRef = (uris: URI[], ref: string): URI | undefined => {
  return uris.find((uri) => {
    // const relativeFsPath =
    //   path.sep + path.relative(getWorkspaceFolder()!.toLowerCase(), uri.fsPath.toLowerCase());

    // if (containsImageExt(ref) || containsOtherKnownExts(ref) || containsUnknownExt(ref)) {
    //   if (isLongRef(ref)) {
    //     return normalizeSlashes(relativeFsPath).endsWith(ref.toLowerCase());
    //   }

    //   const basenameLowerCased = path.basename(uri.fsPath).toLowerCase();

    //   return (
    //     basenameLowerCased === ref.toLowerCase() || basenameLowerCased === `${ref.toLowerCase()}.md`
    //   );
    // }

    // if (isLongRef(ref)) {
    //   return normalizeSlashes(relativeFsPath).endsWith(`${ref.toLowerCase()}.md`);
    // }

    const name = path.parse(uri.fsPath).name.toLowerCase();

    return (
      containsMarkdownExt(path.basename(uri.fsPath)) &&
      name === ref.toLowerCase()
    );
  });
};

export const lineBreakOffsetsByLineIndex = (value: string): number[] => {
  const result = [];
  let index = value.indexOf("\n");

  while (index !== -1) {
    result.push(index + 1);
    index = value.indexOf("\n", index + 1);
  }

  result.push(value.length + 1);

  return result;
};

export const positionToOffset = (
  content: string,
  position: { line: number; column: number }
) => {
  if (position.line < 0) {
    throw new Error("Illegal argument: line must be non-negative");
  }

  if (position.column < 0) {
    throw new Error("Illegal argument: column must be non-negative");
  }

  const lineBreakOffsetsByIndex = lineBreakOffsetsByLineIndex(content);
  if (lineBreakOffsetsByIndex[position.line] !== undefined) {
    return (
      (lineBreakOffsetsByIndex[position.line - 1] || 0) + position.column || 0
    );
  }

  return 0;
};

export const getFileUrlForMarkdownPreview = (filePath: string): string =>
  URI.file(filePath).toString().replace("file://", "");

export const isInFencedCodeBlock = (
  content: string,
  lineNum: number
): boolean => {
  const textBefore = content
    .slice(0, positionToOffset(content, { line: lineNum, column: 0 }))
    .replace(REGEX_FENCED_CODE_BLOCK, "")
    .replace(/<!--[\W\w]+?-->/g, "");
  // So far `textBefore` should contain no valid fenced code block or comment
  return /^( {0,3}|\t)```[^`\r\n]*$[\w\W]*$/gm.test(textBefore);
};

export const trimLeadingSlash = (value: string) =>
  value.replace(/^\/+|^\\+/g, "");
export const normalizeSlashes = (value: string) => value.replace(/\\/gi, "/");

export const fsPathToRef = ({
  path: fsPath,
  keepExt,
  basePath,
}: {
  path: string;
  keepExt?: boolean;
  basePath?: string;
}): string | null => {
  const ref =
    basePath && fsPath.startsWith(basePath)
      ? normalizeSlashes(fsPath.replace(basePath, ""))
      : path.basename(fsPath);

  if (keepExt) {
    return trimLeadingSlash(ref);
  }

  return trimLeadingSlash(
    ref.includes(".") ? ref.slice(0, ref.lastIndexOf(".")) : ref
  );
};

export const parseRef = (rawRef: string): RefT => {
  const dividerPosition = rawRef.indexOf("|");
  if (dividerPosition < 0) {
    return {
      ref: _.trim(rawRef),
      label: _.trim(rawRef),
    };
  } else {
    return {
      ref: _.trim(rawRef.slice(dividerPosition + 1, rawRef.length)),
      label: _.trim(rawRef.slice(0, dividerPosition)),
    };
  }
};

const refRegexp = new RegExp(refPattern, "gi");

export const extractDanglingRefs = (content: string) => {
  const refs: string[] = [];

  content.split(/\r?\n/g).forEach((lineText, _lineNum) => {
    for (const match of matchAll(refRegexp, lineText)) {
      const [, , reference] = match;
      if (reference) {
        // const offset = (match.index || 0) + 2;

        // if (isInFencedCodeBlock(content, lineNum) || isInCodeSpan(content, lineNum, offset)) {
        //   continue;
        // }

        const { ref } = parseRef(reference);

        if (!findUriByRef(getWorkspaceCache().allUris, ref)) {
          refs.push(ref);
        }
      }
    }
  });

  return Array.from(new Set(refs));
};

export const getWorkspaceCache = (): WorkspaceCache => workspaceCache;

export const matchAll = (
  pattern: RegExp,
  text: string
): Array<RegExpMatchArray> => {
  let match: RegExpMatchArray | null;
  const out: RegExpMatchArray[] = [];

  pattern.lastIndex = 0;

  while ((match = pattern.exec(text))) {
    out.push(match);
  }

  return out;
};
