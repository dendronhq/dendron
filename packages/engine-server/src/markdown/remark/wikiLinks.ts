import {
  CONSTANTS,
  DendronError,
  NoteUtils,
  Position,
  VaultUtils,
} from "@dendronhq/common-all";
import { vault2Path } from "@dendronhq/common-server";
import _ from "lodash";
import path from "path";
import { Eat } from "remark-parse";
import Unified, { Plugin } from "unified";
import {
  DendronASTDest,
  DendronASTTypes,
  WikiLinkDataV4,
  WikiLinkNoteV4,
} from "../types";
import { MDUtilsV4 } from "../utils";
import { MDUtilsV5, ProcMode } from "../utilsv5";
import { addError, getNoteOrError, LinkUtils } from "./utils";

export const LINK_REGEX = /^\[\[([^\]\n]+)\]\]/;
/**
 * Does not require wiki link be the start of the word
 */
export const LINK_REGEX_LOOSE = /\[\[([^\]\n]+)\]\]/;

const parseWikiLink = (linkMatch: string) => {
  linkMatch = NoteUtils.normalizeFname(linkMatch);
  return LinkUtils.parseLinkV2(linkMatch);
};

export const matchWikiLink = (text: string) => {
  const match = LINK_REGEX_LOOSE.exec(text);
  if (match) {
    const start = match.index;
    const end = match.index + match[0].length;
    const linkMatch = match[1].trim();
    const link = parseWikiLink(linkMatch);
    return { link, start, end };
  }
  return false;
};

type PluginOpts = CompilerOpts;

type CompilerOpts = {
  convertObsidianLinks?: boolean;
  useId?: boolean;
  prefix?: string;
};

function parseAnchorIfExist(link: string) {
  if (link.indexOf("#") !== -1) {
    return link.split("#");
  } else {
    return [link, undefined];
  }
}

function normalizeSpaces(link: string) {
  return link.replace(/ /g, "%20");
}

const plugin: Plugin<[CompilerOpts?]> = function (
  this: Unified.Processor,
  opts?: PluginOpts
) {
  attachParser(this);
  if (this.Compiler != null) {
    attachCompiler(this, opts);
  }
};

function attachCompiler(proc: Unified.Processor, opts?: CompilerOpts) {
  const copts = _.defaults(opts || {}, {
    convertObsidianLinks: false,
    useId: false,
  });
  const Compiler = proc.Compiler;
  const visitors = Compiler.prototype.visitors;
  if (visitors) {
    visitors.wikiLink = function (node: WikiLinkNoteV4) {
      const pOpts = MDUtilsV5.getProcOpts(proc);
      const data = node.data;
      let value = node.value;

      if (pOpts.mode === ProcMode.NO_DATA) {
        const { alias, anchorHeader } = data;
        const link = value;
        const calias = alias !== value ? `${alias}|` : "";
        const anchor = anchorHeader ? `#${anchorHeader}` : "";
        const vaultPrefix = data.vaultName
          ? `${CONSTANTS.DENDRON_DELIMETER}${data.vaultName}/`
          : "";
        return `[[${calias}${vaultPrefix}${link}${anchor}]]`;
      }

      const { dest } = MDUtilsV4.getDendronData(proc);
      const vault = MDUtilsV4.getVault(proc, data.vaultName);
      // if converting back to dendron md, no further processing
      if (dest === DendronASTDest.MD_DENDRON) {
        return LinkUtils.renderNoteLink({
          link: {
            from: {
              fname: value,
              alias: data.alias,
              anchorHeader: data.anchorHeader,
              vaultName: data.vaultName,
            },
            data: {
              xvault: !_.isUndefined(data.vaultName),
            },
            type: LinkUtils.astType2DLinkType(DendronASTTypes.WIKI_LINK),
            position: node.position as Position,
          },
          dest,
        });
      }

      const { error, engine } = MDUtilsV4.getEngineFromProc(proc);
      if (error) {
        addError(proc, error);
        return "error with engine";
      }

      if (copts.useId && dest === DendronASTDest.HTML) {
        // TODO: check for vault
        const notes = NoteUtils.getNotesByFname({
          fname: value,
          notes: engine.notes,
          vault,
        });
        const { error, note } = getNoteOrError(notes, value);
        if (error) {
          addError(proc, error);
          return "error with link";
        } else {
          value = note!.id;
        }
      }

      switch (dest) {
        case DendronASTDest.MD_REGULAR: {
          const alias = data.alias ? data.alias : value;
          return `[${alias}](${copts.prefix || ""}${normalizeSpaces(value)})`;
        }
        case DendronASTDest.MD_ENHANCED_PREVIEW: {
          const alias = data.alias ? data.alias : value;
          let cleanValue = normalizeSpaces(
            parseAnchorIfExist(value)[0] as string
          );
          if (data.vaultName) {
            const vaultByName = VaultUtils.getVaultByName({
              vaults: engine.vaults,
              vname: data.vaultName,
            });
            if (_.isUndefined(vaultByName)) {
              return `ERROR: ${vaultByName} is not a vault in this workspace`;
            }
            const cpath = vault2Path({
              wsRoot: engine.wsRoot,
              vault: MDUtilsV4.getVault(proc),
            });
            const npath = vault2Path({
              vault: vaultByName!,
              wsRoot: engine.wsRoot,
            });
            cleanValue = path.join(path.relative(cpath, npath), cleanValue);
          }
          return `[${alias}](${copts.prefix || ""}${cleanValue}.md)`;
        }
        case DendronASTDest.HTML: {
          const alias = data.alias ? data.alias : value;
          return `[${alias}](${copts.prefix || ""}${value}.html${
            data.anchorHeader ? "#" + data.anchorHeader : ""
          })`;
        }
        default:
          return `unhandled case: ${dest}`;
      }
    };
  }
}

function attachParser(proc: Unified.Processor) {
  function locator(value: string, fromIndex: number) {
    return value.indexOf("[", fromIndex);
  }

  function parseLink(linkMatch: string) {
    const pOpts = MDUtilsV5.getProcOpts(proc);
    linkMatch = NoteUtils.normalizeFname(linkMatch);
    const out = LinkUtils.parseLinkV2(linkMatch);
    if (_.isNull(out)) {
      throw new DendronError({ message: `link is null: ${linkMatch}` });
    }
    if (pOpts.mode === ProcMode.NO_DATA) {
      return out;
    }

    const procData = MDUtilsV5.getProcData(proc);
    let { vault } = procData;
    const { config, dest, fname } = procData;
    const { engine } = MDUtilsV4.getEngineFromProc(proc);
    if (out.vaultName) {
      const maybeVault = VaultUtils.getVaultByName({
        vaults: engine.vaults,
        vname: out.vaultName,
      });
      if (_.isUndefined(maybeVault)) {
        addError(
          proc,
          new DendronError({
            message: `fname: ${fname}, vault ${
              out.vaultName
            } not found in ${JSON.stringify(out)}`,
          })
        );
      } else {
        vault = maybeVault;
      }
      // default to current note
    }
    if (!out.value) {
      // same file block reference, value is implicitly current file
      out.value = _.trim(NoteUtils.normalizeFname(fname)); // recreate what value (and alias) would have been parsed
      if (!out.alias) out.alias = out.value;
    }
    if (
      dest !== DendronASTDest.MD_DENDRON &&
      config?.useNoteTitleForLink &&
      out.alias === out.value &&
      vault
    ) {
      const wsRoot = engine.wsRoot;
      const note = NoteUtils.getNoteByFnameV5({
        fname: out.value,
        notes: engine.notes,
        vault,
        wsRoot,
      });
      if (note) {
        out.alias = note.title;
      }
    }
    return out;
  }

  function inlineTokenizer(eat: Eat, value: string) {
    const match = LINK_REGEX.exec(value);
    if (match) {
      const linkMatch = match[1].trim();
      try {
        const { value, alias, anchorHeader, vaultName } = parseLink(linkMatch);
        return eat(match[0])({
          type: DendronASTTypes.WIKI_LINK,
          value,
          data: {
            alias,
            anchorHeader,
            vaultName,
          } as WikiLinkDataV4,
        });
      } catch {
        // Broken link, just refuse to parse it
        return;
      }
    }
    return;
  }
  inlineTokenizer.locator = locator;

  const Parser = proc.Parser;
  const inlineTokenizers = Parser.prototype.inlineTokenizers;
  const inlineMethods = Parser.prototype.inlineMethods;
  inlineTokenizers.wikiLink = inlineTokenizer;
  inlineMethods.splice(inlineMethods.indexOf("link"), 0, "wikiLink");
}

export { plugin as wikiLinks };
export { PluginOpts as WikiLinksOpts };
