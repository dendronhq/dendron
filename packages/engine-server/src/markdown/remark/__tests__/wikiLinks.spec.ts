import { DEngineClientV2, WorkspaceOpts } from "@dendronhq/common-all";
import { createLogger } from "@dendronhq/common-server";
import {
  getLogFilePath,
  NoteTestUtilsV4,
  runEngineTestV4,
} from "@dendronhq/common-test-utils";
import _ from "lodash";
import { DendronEngineV2 } from "../../../enginev2";
import { DendronASTData, DendronASTDest } from "../../types";
import { MDUtilsV4 } from "../../utils";
import { wikiLinks, WikiLinksOpts } from "../wikiLinks";

function proc(
  engine: DEngineClientV2,
  dendron: DendronASTData,
  opts?: WikiLinksOpts
) {
  return MDUtilsV4.proc({ engine })
    .data("dendron", dendron)
    .use(wikiLinks, opts);
}

describe("parse", () => {
  let engine: any;
  let dendronData = { dest: DendronASTDest.MD_REGULAR };

  test("basic", () => {
    const resp = proc(engine, dendronData).parse(`[[foo.md]]`);
    expect(resp).toMatchSnapshot();
    // @ts-ignore
    expect(resp.children[0].children[0].type).toEqual("wikiLink");
  });

  test("link with space", () => {
    const resp = proc(engine, dendronData).parse(`[[foo bar]]`);
    expect(resp).toMatchSnapshot();
    // @ts-ignore
    expect(resp.children[0].children[0].type).toEqual("wikiLink");
    // @ts-ignore
    expect(resp.children[0].children[0].value).toEqual("foo bar");
  });

  test("doesn't parse inline code block", () => {
    const resp = proc(engine, dendronData).parse("`[[foo.md]]`");
    expect(resp).toMatchSnapshot("bond");
    // @ts-ignore
    expect(resp.children[0].children[0].type).toEqual("inlineCode");
  });
});

const createEngine = ({ vaults, wsRoot }: WorkspaceOpts) => {
  const logger = createLogger("testLogger", getLogFilePath("engine-server"));
  const engine = DendronEngineV2.createV3({ vaults, wsRoot, logger });
  return engine;
};

const basicSetup = async ({ wsRoot, vaults }: WorkspaceOpts) => {
  await NoteTestUtilsV4.createNote({
    wsRoot,
    fname: "foo",
    vault: vaults[0],
    props: { id: "foo-id" },
  });
};

const getTextString1 = (link = `[[FAQ | dendron.faq]]`) =>
  `See the ${link} for answers for`;
describe("compilev2", () => {
  const linkRegular = "[[foo]]";
  const linkWithAnchor = "[[foo#one]]";
  const linkWithExtension = "[[foo.md]]";
  const linkWithAlias = `[[bar|foo]]`;
  const linkWithSpace = `[[bar|foo bar]]`;
  const textString1 = getTextString1();

  const expected: { [key in DendronASTDest]: any } = {
    [DendronASTDest.MD_DENDRON]: {
      regular: {
        link: linkRegular,
      },
      regularWithAnchor: {
        link: linkWithAnchor,
      },
      regularWithExtension: {
        link: linkRegular,
      },
      alias: {
        link: linkWithAlias,
      },
      withId: {
        link: linkRegular,
      },
      withSpace: {
        link: linkWithSpace,
      },
      textString1: {
        link: getTextString1(`[[FAQ|dendron.faq]]`),
      },
    },
    [DendronASTDest.HTML]: {
      regular: {
        link: `[foo](foo.html)`,
      },
      regularWithAnchor: {
        link: `[foo](foo.html#one)`,
      },
      regularWithExtension: {
        link: `[foo](foo.html)`,
      },
      alias: {
        link: `[bar](foo.html)`,
      },
      withId: {
        link: `[foo](foo-id.html)`,
      },
      withSpace: {
        link: `[bar](foo bar.html)`,
      },
      textString1: {
        link: getTextString1(`[FAQ](dendron.faq.html)`),
      },
    },
    [DendronASTDest.MD_REGULAR]: {
      regular: {
        link: `[foo](foo)`,
      },
      regularWithAnchor: {
        link: `[foo](foo)`,
      },
      regularWithExtension: {
        link: `[foo](foo)`,
      },
      alias: {
        link: `[bar](foo)`,
      },
      withId: {
        link: `[foo](foo-id)`,
      },
      withSpace: {
        link: `[bar](foo%20bar)`,
      },
      textString1: {
        link: getTextString1(`[FAQ](dendron.faq)`),
      },
    },
    [DendronASTDest.MD_ENHANCED_PREVIEW]: {
      regular: {
        link: `[foo](foo.md)`,
      },
      regularWithAnchor: {
        link: `[foo](foo.md)`,
      },
      regularWithExtension: {
        link: `[foo](foo.md)`,
      },
      alias: {
        link: `[bar](foo.md)`,
      },
      withId: {
        link: `[foo](foo-id.md)`,
      },
      withSpace: {
        link: `[bar](foo%20bar.md)`,
      },
      textString1: {
        link: getTextString1(`[FAQ](dendron.faq.md)`),
      },
    },
  };

  const testCases = [
    {
      testCase: "regular",
      linkProcess: linkRegular,
      preSetupHook: undefined,
      procOpts: {},
    },
    {
      testCase: "regularWithAnchor",
      linkProcess: linkWithAnchor,
      preSetupHook: undefined,
      procOpts: {},
    },
    {
      testCase: "regularWithExtension",
      linkProcess: linkWithExtension,
      preSetupHook: undefined,
      procOpts: {},
    },
    { testCase: "alias", linkProcess: linkWithAlias },
    {
      testCase: "withId",
      linkProcess: linkWithExtension,
      preSetupHook: basicSetup,
      procOpts: { useId: true },
    },
    {
      testCase: "textString1",
      linkProcess: textString1,
    },
    {
      testCase: "withSpace",
      linkProcess: linkWithSpace,
      preSetupHook: basicSetup,
    },
  ];

  Object.values(DendronASTDest).map((key) => {
    const dest = key as DendronASTDest;
    describe(dest, () => {
      test.each(testCases.map((ent) => [ent.testCase, ent]))(
        "%p",
        async (_key, obj: typeof testCases[0]) => {
          const { testCase, linkProcess, preSetupHook, procOpts } = obj;
          await runEngineTestV4(
            async ({ engine }) => {
              const resp = await proc(
                engine,
                {
                  dest,
                },
                procOpts
              ).process(linkProcess);
              expect(resp).toMatchSnapshot();
              expect(_.trim(resp.toString())).toEqual(
                expected[dest][testCase].link
              );
            },
            { expect, createEngine, preSetupHook }
          );
        }
      );
    });
  });
});
