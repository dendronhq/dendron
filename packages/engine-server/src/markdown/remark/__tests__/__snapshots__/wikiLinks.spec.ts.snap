// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compilev2 HTML "alias" 1`] = `
VFile {
  "contents": "[bar](foo.html)
",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`compilev2 HTML "regular" 1`] = `
VFile {
  "contents": "[foo](foo.html)
",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`compilev2 HTML "regularWithAnchor" 1`] = `
VFile {
  "contents": "[foo](foo.html#one)
",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`compilev2 HTML "regularWithExtension" 1`] = `
VFile {
  "contents": "[foo](foo.html)
",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`compilev2 HTML "textString1" 1`] = `
VFile {
  "contents": "See the [FAQ](dendron.faq.html) for answers for
",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`compilev2 HTML "withId" 1`] = `
VFile {
  "contents": "[foo](foo-id.html)
",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`compilev2 HTML "withSpace" 1`] = `
VFile {
  "contents": "[bar](foo bar.html)
",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`compilev2 MD_DENDRON "alias" 1`] = `
VFile {
  "contents": "[[bar|foo]]
",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`compilev2 MD_DENDRON "regular" 1`] = `
VFile {
  "contents": "[[foo]]
",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`compilev2 MD_DENDRON "regularWithAnchor" 1`] = `
VFile {
  "contents": "[[foo#one]]
",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`compilev2 MD_DENDRON "regularWithExtension" 1`] = `
VFile {
  "contents": "[[foo]]
",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`compilev2 MD_DENDRON "textString1" 1`] = `
VFile {
  "contents": "See the [[FAQ|dendron.faq]] for answers for
",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`compilev2 MD_DENDRON "withId" 1`] = `
VFile {
  "contents": "[[foo]]
",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`compilev2 MD_DENDRON "withSpace" 1`] = `
VFile {
  "contents": "[[bar|foo bar]]
",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`compilev2 MD_ENHANCED_PREVIEW "alias" 1`] = `
VFile {
  "contents": "[bar](foo.md)
",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`compilev2 MD_ENHANCED_PREVIEW "regular" 1`] = `
VFile {
  "contents": "[foo](foo.md)
",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`compilev2 MD_ENHANCED_PREVIEW "regularWithAnchor" 1`] = `
VFile {
  "contents": "[foo](foo.md)
",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`compilev2 MD_ENHANCED_PREVIEW "regularWithExtension" 1`] = `
VFile {
  "contents": "[foo](foo.md)
",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`compilev2 MD_ENHANCED_PREVIEW "textString1" 1`] = `
VFile {
  "contents": "See the [FAQ](dendron.faq.md) for answers for
",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`compilev2 MD_ENHANCED_PREVIEW "withId" 1`] = `
VFile {
  "contents": "[foo](foo-id.md)
",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`compilev2 MD_ENHANCED_PREVIEW "withSpace" 1`] = `
VFile {
  "contents": "[bar](foo%20bar.md)
",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`compilev2 MD_REGULAR "alias" 1`] = `
VFile {
  "contents": "[bar](foo)
",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`compilev2 MD_REGULAR "regular" 1`] = `
VFile {
  "contents": "[foo](foo)
",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`compilev2 MD_REGULAR "regularWithAnchor" 1`] = `
VFile {
  "contents": "[foo](foo)
",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`compilev2 MD_REGULAR "regularWithExtension" 1`] = `
VFile {
  "contents": "[foo](foo)
",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`compilev2 MD_REGULAR "textString1" 1`] = `
VFile {
  "contents": "See the [FAQ](dendron.faq) for answers for
",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`compilev2 MD_REGULAR "withId" 1`] = `
VFile {
  "contents": "[foo](foo-id)
",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`compilev2 MD_REGULAR "withSpace" 1`] = `
VFile {
  "contents": "[bar](foo%20bar)
",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`parse basic 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "data": Object {
            "alias": "foo",
            "anchorHeader": undefined,
          },
          "position": Position {
            "end": Object {
              "column": 11,
              "line": 1,
              "offset": 10,
            },
            "indent": Array [],
            "start": Object {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "wikiLink",
          "value": "foo",
        },
      ],
      "position": Position {
        "end": Object {
          "column": 11,
          "line": 1,
          "offset": 10,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": Object {
    "end": Object {
      "column": 11,
      "line": 1,
      "offset": 10,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`parse doesn't parse inline code block: bond 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "position": Position {
            "end": Object {
              "column": 13,
              "line": 1,
              "offset": 12,
            },
            "indent": Array [],
            "start": Object {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "inlineCode",
          "value": "[[foo.md]]",
        },
      ],
      "position": Position {
        "end": Object {
          "column": 13,
          "line": 1,
          "offset": 12,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": Object {
    "end": Object {
      "column": 13,
      "line": 1,
      "offset": 12,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`parse link with space 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "data": Object {
            "alias": "foo bar",
            "anchorHeader": undefined,
          },
          "position": Position {
            "end": Object {
              "column": 12,
              "line": 1,
              "offset": 11,
            },
            "indent": Array [],
            "start": Object {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "wikiLink",
          "value": "foo bar",
        },
      ],
      "position": Position {
        "end": Object {
          "column": 12,
          "line": 1,
          "offset": 11,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": Object {
    "end": Object {
      "column": 12,
      "line": 1,
      "offset": 11,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;
